### ISSUE TITLE
Removing VK_MVK_moltenvk extension breaks existing applications at runtime

### ISSUE THREAD

**Issue by SRSaunders** [KhronosGroup/MoltenVK#1884]
Removing the **VK_MVK_moltenvk** extension breaks existing applications that previously required it to gain access to `vkGetMoltenVKConfigurationMVK() `and `vkSetMoltenVKConfigurationMVK()`

Applications that used the config API via **VK_MVK_moltenvk** extension the will now fail at runtime with the following error:
_Sys_Error: Cannot create a Vulkan instance because the following required extension(s) are not supported: VK_MVK_moltenvk_

In addition, moving to the new approach (via **mvk_config.h** without the extension) will not be compatible with older runtime versions of the library.  I presume developers may now have to test for release versions and include code for both approaches in their applications.  A bit messy.

While I understand the need to move forward, isn't there a way to mark this extension as deprecated for a period of time to allow developers and end users to adapt?

**Comment by billhollings** [comment]
Thanks for the feedback. We definitely want this to be as backwards compatible as possible.

> Sys_Error: Cannot create a Vulkan instance because the following required extension(s) are not supported: VK_MVK_moltenvk

What's the source of that error? Is it an app or library checking for the existence of  the `VK_MVK_moltenvk` extension before attempting to call the config functions (which are still available)?

>In addition, moving to the new approach (via mvk_config.h without the extension) will not be compatible with older runtime versions of the library. I presume developers may now have to test for release versions and include code for both approaches in their applications. A bit messy, but leaving vk_mvk_moltenvk.h around for a while will help.

Yes. The intention of leaving `vk_mvk_moltenvk.h` in place, and including everything it used to, is to allow existing apps to continue to connect they way they always have.

> isn't there a way to mark this extension as deprecated for a period of time to allow developers and end users to adapt?

There really isn't a runtime way to mark an extension as deprecated. But we can simply reinstate it to avoid the regression errors, and continue to document that it is deprecated. The functions that are deprecated will also alert app devs to that fact.

In summary...from your points, it looks like re-advertising the `VK_MVK_moltenvk` might be sufficient to fix the problems you've encountered or are anticipating?

**Comment by SRSaunders** [comment]
Thanks for your quick response.

> What's the source of that error? Is it coming from an app or library that is checking for the existence of the VK_MVK_moltenvk extension before attempting to call the config functions (which are still available and undeprecated)?

Yes, you are correct.  The application (RBDoom3-BFG) is checking for the existence of `VK_MVK_moltenvk` before calling the config functions (which are defined in **vk_mvk_moltenvk.h**, which still exists).  I thought (perhaps sometime in the past) this extension was required to use the config functions at runtime - so I am responsible for that code, good or bad :)

However, I just did a quick test against the 1.3.243.0 version of the SDK, and the extension is indeed not required even before your new change.  The config functions can be called without error.  So perhaps this is just my misunderstanding and I have a leftover piece of code asking for the extension when it hasn't been necessary all along.  Would you please confirm if this was a mistake on my part from the beginning, or if something has changed along the path to today.

**Comment by billhollings** [comment]
> The application (RBDoom3-BFG) is checking for the existence of VK_MVK_moltenvk before calling the config functions

> However, I just did a quick test against the 1.3.243.0 version of the SDK, and the extension is indeed not required even before your new change. The config functions can be called without error. So perhaps this is just my misunderstanding and I have a leftover piece of code asking for the extension when it hasn't been necessary all along.

MoltenVK has always exposed those functions even without the extension being enabled. But you have not made a mistake. It is Vulkan best-practise to verify extensions before calling functions. And if you're doing it, others will be too!

Thanks for confirming. I'll issue another PR to re-advertise the extension.

**Comment by SRSaunders** [comment]
Thanks very much.

**Comment by billhollings** [comment]
PR #1886 should fix this. Please retest with latest MoltenVK, and close this issue if it is fixed.

**Comment by SRSaunders** [comment]
PR #1886 works perfectly.  I tested it using static linking to MoltenVK (debug build) and dynamic linking (release build).

I see new compile-time deprecation warnings for `vkGetMoltenVKConfigurationMVK()` and `vkSetMoltenVKConfigurationMVK()`, as well as a runtime deprecation warning for `VK_MVK_moltenvk`.  The messages are clear and help with making changes going forward.  And most importantly, the existing app runs without breakage.

Thanks very much, I think this solves it!

**Comment by billhollings** [comment]
Thanks for all your testing!
